### PR TITLE
Nazar/feature/set cases

### PR DESCRIPTION
**Pull request by Nazar-kob**

### CODE REVIEW

**Review by StoliarskyiKostiantyn — APPROVED** [review]

File: app/controllers/s3_bucket.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        self.bucket_name = app.config['AWS_BUCKET_NAME']
+        self.allowed_extensions = app.config['ALLOWED_EXTENSIONS']
+        self.s3 = boto3.client(
+            "s3",
+            aws_access_key_id=app.config['AWS_ACCESS_KEY'],
+            aws_secret_access_key=app.config['AWS_SECRET_ACCESS_KEY']
+        )
+
+    def _generate_img_uid(self):
+        return str(uuid.uuid4())[:8]
```
Comment (nixissmtp) [inline app/controllers/s3_bucket.py]: return to 32 characters

File: app/controllers/s3_bucket.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    def upload_cases_imgs(self, file, case_name, img_type='title') -> tuple[bool, str]:
```
Comment (nixissmtp) [inline app/controllers/s3_bucket.py]: add Type hints

Comment (nixissmtp) [reply app/controllers/s3_bucket.py]: replace filestorage by file from python (io)

File: app/controllers/s3_bucket.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+        extension_files = file.filename.split(".")[-1]
+
+        if extension_files not in self.allowed_extensions:
```
Comment (nixissmtp) [inline app/controllers/s3_bucket.py]: add filetype library: https://pypi.org/project/filetype/

File: app/templates/case/add.html [hunk 4]
```diff
@@ -13,24 +14,48 @@ <h3 class="text-xl font-semibold text-gray-900 dark:text-white"> Add new case </
...
+            <input type="text" name="sub_title"  class="shadow-sm bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-600 focus:border-blue-600 block w-full p-2.5 dark:bg-gray-600 dark:border-gray-500 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500" placeholder="Sub title" required />
           </div>
           <div class="col-span-6 sm:col-span-3">
-            <label for="password" class="block mb-2 text-sm font-medium text-gray-900 dark:text-white" >Password</label >
-            <input type="password" name="password" id="user-add-password" class="shadow-sm bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-600 focus:border-blue-600 block w-full p-2.5 dark:bg-gray-600 dark:border-gray-500 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500" required />
+            <label class="block mb-2 text-sm font-medium text-gray-900 dark:text-white" >Upload main image</label >
+            <input type="file" name="title_image" required>
```
Comment (nixissmtp) [inline app/templates/case/add.html]: add filetypes

File: app/templates/case/add.html [hunk 5]
```diff
@@ -13,24 +14,48 @@ <h3 class="text-xl font-semibold text-gray-900 dark:text-white"> Add new case </
...
           </div>
           <div class="col-span-6 sm:col-span-3">
-            <label for="password_confirmation" class="block mb-2 text-sm font-medium text-gray-900 dark:text-white" >Confirm Password</label >
-            <input type="password" name="password_confirmation" id="user-add-password_confirmation" class="shadow-sm bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-600 focus:border-blue-600 block w-full p-2.5 dark:bg-gray-600 dark:border-gray-500 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500" required />
+            <label class="block mb-2 text-sm font-medium text-gray-900 dark:text-white" >Upload sub main image</label >
+            <input type="file" name="sub_title_image" required>
```
Comment (nixissmtp) [inline app/templates/case/add.html]: add filetypes

File: app/templates/case/add.html [hunk 6]
```diff
@@ -13,24 +14,48 @@ <h3 class="text-xl font-semibold text-gray-900 dark:text-white"> Add new case </
...
-            <input type="checkbox" name="activated" id="user-add-activated" class="shadow-sm bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-600 focus:border-blue-600 block w-full p-2.5 dark:bg-gray-600 dark:border-gray-500 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500" required />
+            <label for="role" class="block mb-2 text-sm font-medium text-gray-900 dark:text-white" >Stacks</label >
+            <div class="max-h-24 p-2.5 overflow-y-auto shadow-sm bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-600 focus:border-blue-600 block w-full dark:bg-gray-600 dark:border-gray-500 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500">
+              {{ form.stacks }}
+            </div>
+          </div>
+
+          <div class="col-span-6 sm:col-span-3">
+            <label  class="block mb-2 text-sm font-medium text-gray-900 dark:text-white" >Upload sub images</label >
+            <input type="file" name="sub_images" multiple >
```
Comment (nixissmtp) [inline app/templates/case/add.html]: add filetypes validation

File: app/views/case.py [hunk 7]
```diff
@@ -22,89 +23,121 @@
...
+        session.add(new_case)
+        session.commit()
+        session.refresh(new_case)
+
+        for id in form.stacks.data:
+            new_stack = m.CaseStack(case_id=new_case.id, stack_id=int(id))
+            session.add(new_stack)
+        session.commit()
+
+        if form.sub_images.data:
```
Comment (nixissmtp) [inline app/views/case.py]: remove if

File: app/views/case.py [hunk 8]
```diff
@@ -22,89 +23,121 @@
...
-            email=form.email.data,
-            password=form.password.data,
-            activated=form.activated.data,
-        )
-        log(log.INFO, "Form submitted. User: [%s]", user)
-        flash("User added!", "success")
-        user.save()
-        return redirect(url_for("user.get_all"))
+def update(id: int):
+    u = db.session.scalar(m.Case.select().where(m.Case.id == id))
```
Comment (nixissmtp) [inline app/views/case.py]: replace u by case

File: app/views/case.py [hunk 9]
```diff
@@ -22,89 +23,121 @@
...
+    log(log.INFO, "Case updated. Case: [%s]", u)
+    flash("Case updated!", "success")
+    return "ok", 200
 
 
 @bp.route("/delete/<int:id>", methods=["DELETE"])
 @login_required
 def delete(id: int):
-    u = db.session.scalar(m.SuperUser.select().where(m.SuperUser.id == id))
+    u = db.session.scalar(m.Case.select().where(m.Case.id == id))
```
Comment (nixissmtp) [inline app/views/case.py]: replace by session.get

File: pyproject.toml [hunk 10]
```diff
@@ -19,6 +19,7 @@ pydantic = "^1.10.7"
 alchemical = "^0.7.1"
 passlib = "^1.7.4"
 bcrypt = "^4.0.1"
+boto3 = "^1.28.11"
```
Comment (nixissmtp) [inline pyproject.toml:22]: add https://pypi.org/project/boto3-stubs/

**Review by nixissmtp — CHANGES_REQUESTED** [review]

File: app/controllers/s3_bucket.py (hunk 1 above)
Comment (Nazar-kob) [reply app/controllers/s3_bucket.py]: done

**Pull request merged**